File: 01-dmoj-dwite09c3p2/program.c
#include <stdio.h>
int main (void) {
    int n;
    // setting the number of fibonacci numbers to be computed capped at 45
    // as F(45) is  1,134,903,170 which is just over the bounds of 10^9 so this 
    // will cover the numbers between (0 <= n <= 10^9) as per problem spec
    int fib[46];
    // the base values of the fibbonaci numbers
    fib[0] = 0;
    fib[1] = 1;
    // precomputing the fib numbers fromm 0 - 45
    for (int i = 2; i < 46; ++i) {
        fib[i] = fib[i-1] + fib[i-2];
    }

    // processing 5 inputs at a time
    for (int i = 0; i < 5; ++i) {
        scanf("%d", &n);
        int high_idx = 0;
        //find the smallest fib number that is >= n, looping through to find the 
        // first fib number that is big enough to reach or pass n
        while (high_idx < 46 && fib[high_idx] < n) {
            high_idx ++;
        }
        int high = fib[high_idx];
        //finding the largest fib number that is <= n
        // same pattern as the loop above but this time starting at the largest value
        // working down to find the first fib which is smaller than n
        int low_idx = 45;
        while (low_idx >= 0 && fib[low_idx] > n) {
            low_idx --;
        }
        int low = fib[low_idx];
        // n is now between the two bounds that we set earlier
        // comparing the two distances and choosing which one is the smaller one
        // ties will automatically go to the larger fib number which is why we use the
        // < instead of <= to avoid ties
        // if the given number is a fib number itself (low == high) then it will just 
        // pprint the high which is n itself
        if (n - low < high - n) {
            printf("%d\n", low);
        } else {
            printf("%d\n", high);
        }
    }

    return 0;
}

File: 01-dmoj-dwite09c3p2/program.py
# The input bound is 10^9,
# so we calculate Fibonacci numbers up until the first one that exceeds the bound.
fibonacci = [0, 1]
while fibonacci[-1] < 10 ** 9:
    fibonacci.append(fibonacci[-2] + fibonacci[-1])

for _ in range(5):
    n = int(input())
    # `low` is the largest Fibonacci number less than or equal to `n`.
    # `high` is the smallest Fibonacci number greater than or equal to `n`.
    low, high = 0, 0
    for fib_num in fibonacci:
        # If we are searching in ascending order,
        # the last Fibonacci number less than or equal to `n` is `low`,
        # and the first Fibonacci number greater than or equal to `n` is `high`.
        if fib_num <= n:
            low = fib_num
        if high == 0 and fib_num >= n:
            high = fib_num
    # Print the lower Fibonacci number only if its difference from `n` is less than that of the higher one.
    print(low if n - low < high - n else high)
